### PR TITLE
Integrate: Alerts

### PR DESCRIPTION
**Pull request by Rijidij**
Here is an example of a page displaying multiple alerts:
Screenshot from 2021-07-02 11-13-49

### CODE REVIEW

File: etna/alerts/templatetags/alert_tags.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    except AttributeError:
+        pass
+
+    if alert and alert.active:
+        return alert
+    else:
+        return None
+
+
+@register.inclusion_tag('alerts/tags/alerts.html', takes_context=False)
```
Comment (danbentley) [inline etna/alerts/templatetags/alert_tags.py:28]: `takes_context` defaults to `None` so you should be able to remove this.

File: etna/alerts/templatetags/alert_tags.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    # Get alert for current page, if any.
+    alert = get_page_alert(page)
+    if alert:
+        alerts.insert(0, alert)
+
+    # Get current page ancestors.
+    branch = Page.objects.ancestor_of(page).specific()
+    # Iterate backwards and, if page has alert and "cascade" is true,
+    # add to alerts list.
+    for i in reversed(range(branch.count())):
```
Comment (danbentley) [inline etna/alerts/templatetags/alert_tags.py:44]: A more idiomatic way to loop through the pages in reverse order is:

```
for page in Page.objects.ancestor_of(page).specific().reverse():
    …
```

Is there a reason why we need to gather the alerts in reverse order but insert them at the front of the list?

File: etna/home/models.py [hunk 3]
```diff
@@ -1,7 +1,18 @@
 from django.db import models
 
 from wagtail.core.models import Page
+from wagtail.snippets.edit_handlers import SnippetChooserPanel
 
 
 class HomePage(Page):
-    pass
+    alert = models.ForeignKey(
```
Comment (danbentley) [inline etna/home/models.py:8]: I can see the need to add the alerts to a lot of pages. Have you considered using a mixin?

File: etna/text_formats/blocks.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from wagtail.core import blocks
```
Comment (danbentley) [inline etna/text_formats/blocks.py:1]: I really like the idea of using a separate app for rich text fields, it helps reduce the need for the dreaded `base` app too!

I think the name `text_formats` might be a bit restrictive though. In the past, we've written rich text converters, editor features as well as supporting [legacy rich text](https://docs.wagtail.io/en/v2.12.4/reference/contrib/legacy_richtext.html) and it'd be nice to keep this all in one place.

Wagtail is a bit inconsistent with its naming here, using both [`rich_text`](https://github.com/wagtail/wagtail/tree/main/wagtail/admin/rich_text ) and [`richtext`](https://github.com/wagtail/wagtail/tree/main/wagtail/contrib/legacy/richtext) but something along those lines would be good.

**Review by danbentley — CHANGES_REQUESTED** [review]

File: sass/includes/_alert-message.scss [hunk 5]
```diff
@@ -25,13 +25,19 @@
...
     &__description {
         margin-bottom: 0;
         display: inline-block;
     }
 
+    &__description a {
+        color: $color__link-hover;
+    }
+
     &__link {
```
Comment (AshTNA) [inline sass/includes/_alert-message.scss:41]: Would you be able to remove this `&__link` class, as it is not used if the only way to access the links  is via the `<a>` tag itself

File: sass/includes/_alert-message.scss [hunk 6]
```diff
@@ -25,13 +25,19 @@
...
+        font-weight: bold;
+        text-transform: uppercase;
     }
 
     &__description {
         margin-bottom: 0;
         display: inline-block;
     }
 
+    &__description a {
```
Comment (AshTNA) [inline sass/includes/_alert-message.scss:37]: Would you be able to remove this, as the styling  for the `<a>` should be nested inside main `&__description` class.

File: sass/includes/_alert-message.scss [hunk 7]
```diff
@@ -25,13 +25,19 @@
         padding-left: 0.4rem;
         padding-right: 0.4rem;
         margin-right: 0.4rem;
+        font-weight: bold;
+        text-transform: uppercase;
     }
 
     &__description {
```
Comment (AshTNA) [inline sass/includes/_alert-message.scss:32]: Could you please have the &__description class as so:

```SASS
    &__description {
        margin-bottom: 0;
        display: inline-block;
        a {
            color: $color__link-hover;
        }
        p {
            margin-bottom: 0;
        }
    }
```

This nests the existing `<a>` class you added within the parent `&__description` block, and also adds a `margin-bottom: 0`  to the `<p>` which we weren't expecting Wagtail to render in our static version.

**Review by AshTNA — CHANGES_REQUESTED** [review]
Hi @Rijidij,

Would you be able make some changes to the SASS? This makes it a bit cleaner and accounts for the `<p>` tag we weren't expecting to be rendered.

Thanks :+1:

File: sass/includes/_alert-message.scss (hunk 5 above)
Comment (Rijidij) [reply sass/includes/_alert-message.scss:41]: No worries.

File: etna/text_formats/blocks.py (hunk 4 above)
Comment (Rijidij) [reply etna/text_formats/blocks.py]: I've renamed the app `richtexts`, to contain any variations that may be required.

**Review by AshTNA — APPROVED** [review]
All good on the frontend :+1:

File: etna/alerts/templatetags/alert_tags.py (hunk 2 above)
Comment (Rijidij) [reply etna/alerts/templatetags/alert_tags.py]: This code has been simplified. The list reversal is no longer required to return pages in hierarchical order.

File: etna/home/models.py (hunk 3 above)
Comment (Rijidij) [reply etna/home/models.py]: A mixin has been created for the alert field and added to the page models.

File: etna/text_formats/blocks.py (hunk 4 above)
Comment (Rijidij) [reply etna/text_formats/blocks.py]: I've removed the richtext(s) app as it's not needed using the features defined in settings as we discussed.
I might need to add it back in later for blocks requiring different feature sets. But I will define those sets in settings as well.

**Comment by danbentley** [comment]
Approved. Excellent work!

**Review by danbentley — APPROVED** [review]

**Comment by danbentley** [comment]
I've rebased this work onto `main` and removed some commits that were causing conflicts but were later reverted in the timeline. 

Additionally, I've consolidated the migrations, manually added the `AlertMixin` to explorer pages (there were too many changes in `collections/models.py` to do it any other way), moved alerts to the settings tab and fixed the unit test.

@Rijidij would you mind checking out the branch and giving it a quick test?

File: etna/collections/models.py [hunk 8]
```diff
@@ -155,6 +159,7 @@ class ResultsPage(TeaserImageMixin, Page):
         FieldPanel("introduction"),
     ]
     promote_panels = Page.promote_panels + TeaserImageMixin.promote_panels
+    settings_panels = Page.settings_panels + AlertMixin.settings_panels
 
     content_panels = [
```
Comment (Rijidij) [inline etna/collections/models.py:164]: This is the issue I found from the first merge.
content_panels is previously assigned. Line 154/158

Comment (Rijidij) [reply etna/collections/models.py:164]: Otherwise, all OK.

**Review by gtvj — APPROVED** [review]

File: etna/collections/models.py (hunk 8 above)
Comment (danbentley) [reply etna/collections/models.py:164]: Thanks for checking.

I noticed that too. It's not from a merge but something that was introduced by mistake when I introduced the CIIM client work.

I've fixed in my explorer integration branch.

**Review by danbentley — APPROVED** [review]

**Pull request merged**